crypto/crmf/crmf_pbm.c

```diff
@@ -0,0 +1,220 @@
+                      const unsigned char *msg, size_t msglen,
+                      const unsigned char *sec, size_t seclen,
+                      unsigned char **mac, unsigned int *maclen)
+{
+    int mac_nid, hmac_md_nid = NID_undef;
+    const EVP_MD *m = NULL;
+    EVP_MD_CTX *ctx = NULL;
+    unsigned char basekey[EVP_MAX_MD_SIZE];
+    unsigned int bklen = EVP_MAX_MD_SIZE;
+    int64_t iterations;
+    unsigned char *mac_res = 0;
+    int ok = 0;
+
+    if (mac == NULL || pbmp == NULL || pbmp->mac == NULL
+        || pbmp->mac->algorithm == NULL || msg == NULL || sec == NULL) {
+        CRMFerr(CRMF_F_OSSL_CRMF_PBM_NEW, CRMF_R_NULL_ARGUMENT);
+        goto err;
+    }
+    if ((mac_res = OPENSSL_malloc(EVP_MAX_MD_SIZE)) == NULL) {
+        CRMFerr(CRMF_F_OSSL_CRMF_PBM_NEW, ERR_R_MALLOC_FAILURE);
+        goto err;
+    }
+
+    /*
+     * owf identifies the hash algorithm and associated parameters used to
+     * compute the key used in the MAC process.  All implementations MUST
+     * support SHA-1.
+     */
+    if ((m = EVP_get_digestbyobj(pbmp->owf->algorithm)) == NULL) {
+        CRMFerr(CRMF_F_OSSL_CRMF_PBM_NEW, CRMF_R_UNSUPPORTED_ALGORITHM);
+        goto err;
+    }
+
+    if ((ctx = EVP_MD_CTX_new()) == NULL) {
+        CRMFerr(CRMF_F_OSSL_CRMF_PBM_NEW, ERR_R_MALLOC_FAILURE);
+        goto err;
+    }
+
+    /* compute the basekey of the salted secret */
+    if (!(EVP_DigestInit_ex(ctx, m, NULL)))
+        goto err;
+    /* first the secret */
+    if (!EVP_DigestUpdate(ctx, sec, seclen))
+        goto err;
+    /* then the salt */
+    if (!EVP_DigestUpdate(ctx, pbmp->salt->data, pbmp->salt->length))
+        goto err;
+    if (!(EVP_DigestFinal_ex(ctx, basekey, &bklen)))
+        goto err;
+    if (!ASN1_INTEGER_get_int64(&iterations, pbmp->iterationCount)
+        || iterations < 100 /* min from RFC */
+        || iterations > OSSL_CRMF_PBM_MAX_ITERATION_COUNT) {
+        CRMFerr(CRMF_F_OSSL_CRMF_PBM_NEW, CRMF_R_BAD_PBM_ITERATIONCOUNT);
+        goto err;
+    }
+
+    /* the first iteration was already done above */
+    while (--iterations > 0) {
+        if (!(EVP_DigestInit_ex(ctx, m, NULL)))
+            goto err;
+        if (!EVP_DigestUpdate(ctx, basekey, bklen))
+            goto err;
+        if (!(EVP_DigestFinal_ex(ctx, basekey, &bklen)))
+            goto err;
+    }
+
+    /*-
+     * mac identifies the algorithm and associated parameters of the MAC
+     * function to be used.  All implementations MUST support HMAC-SHA1 [HMAC].
+     * All implementations SHOULD support DES-MAC and Triple-DES-MAC [PKCS11].
+     */
+    mac_nid = OBJ_obj2nid(pbmp->mac->algorithm);
+
+    if (!EVP_PBE_find(EVP_PBE_TYPE_PRF, mac_nid, NULL, &hmac_md_nid, NULL)
+        || ((m = EVP_get_digestbynid(hmac_md_nid)) == NULL)) {
+        CRMFerr(CRMF_F_OSSL_CRMF_PBM_NEW, CRMF_R_UNSUPPORTED_ALGORITHM);
+        goto err;
+    }
+    if (HMAC(m, basekey, bklen, msg, msglen, mac_res, maclen) != NULL)
```

Comment:
Can we use EVP_MAC here?

Reply:
Yes we can - although it's not obvious what the instant benefit might be.

On it's way, pending review https://github.com/mpeylo/cmpossl/pull/163

Reply:
Done, taking over @mpeylo's code.

Reply:
> Yes we can - although it's not obvious what the instant benefit might be.

That API will be deprecated in OpenSSL 3. The EVP APIs will benefit from the new Provider architecture (https://www.openssl.org/docs/OpenSSL300Design.html)...so for example, a user has the ability to mandate that all crypto goes through the new FIPS module.